### ISSUE TITLE
_Nelementsof / _Lengthof / _Countof operator (C2y)

### ISSUE THREAD

**Issue by alejandro-colomar** [cc65/cc65#2530]
Hi,

An operator for getting the number of elements of an array was recently added to C2y, (I proposed it, and also have patches ready for GCC).  We're still in a discussion about the name, trying to change what WG14 (the standard committee) voted, so you might be interested in it:

<https://inbox.sourceware.org/gcc/br6lhebpiivavhzdyldsmbpheldcrnpx6zqcxhxwrtohxeqs2m@of3ra3mahmew/T/#m2a4fd73ac91f4b966fd9243f7f0c62e78844b1a2>

Cheers,
Alex

**Comment by alejandro-colomar** [comment]
On the other hand, I only see support for C89 and C99, so you may be not interested in this operator.  (Or you may want to provide it as an extension?)

**Comment by GorillaSapiens** [comment]
as i understand, _Countof is an alias for _Nelementsof , which is just the classic (sizeof(foo) / sizeof(foo[0])) trick.  couldn't they be done with a macro?  the only new / useful one is _Lengthof which is the same EXCEPT when used on a string, in which case it's the length excluding the null terminator.

**Comment by alejandro-colomar** [comment]
> as i understand, _Countof is an alias for _Nelementsof ,

No, the name was under discussion in the committee, and we ended up calling the operator `_Countof`.  `_Nelementsof` doesn't exist.

<https://www.open-std.org/jtc1/sc22/wg14/www/docs/n3550.pdf#subsubsection.0.6.5.4.5>
<https://thephd.dev/the-big-array-size-survey-for-c>
<https://thephd.dev/the-big-array-size-survey-for-c-results>

> which is just the classic (sizeof(foo) / sizeof(foo[0])) trick. couldn't they be done with a macro?

There are things you can do with the new operator that you couldn't do with the macro:

-  `_Countof` can count arrays of objects whose size is 0 (this is non-standard, but supported by some implementations)

```c
alx@debian:~/tmp$ cat cnt.c 
#include <stdio.h>
int
main(void)
{
	int  a[42][0];

	printf("%zu\n", _Countof(a));
	printf("%zu\n", sizeof(a) / sizeof(*a));
}
```
```c
alx@debian:~/tmp$ gcc cnt.c 
cnt.c: In function ‘main’:
cnt.c:8:35: warning: division by zero [-Wdiv-by-zero]
    8 |         printf("%zu\n", sizeof(a) / sizeof(*a));
      |                                   ^
alx@debian:~/tmp$ ./a.out 
42
Floating point exception
```

-  `_Countof` protects you from passing pointers (it is a constraint violation).

-  I'm working on an extension that will allow `_Countof` working for array parameters to functions, and return the number of elements declared in the prototype (as if it were a real function).

> the only new / useful one is _Lengthof which is the same EXCEPT when used on a string, in which case it's the length excluding the null terminator.

No, there's no operator that is an overload between `_Countof` and `strlen(3)`, and I'd find it quite dangerous.

**Comment by GorillaSapiens** [comment]
looks like you've retconned the name of the issue here, and your ask.
sounds like the standard isn't settled yet, in which case i'd be firmly against taking any action at this time.

what is the use of an int[42][0] ? seems like this use case was concocted to justify its own existence.

not sure what is "quite dangerous" about a compile time macro giving the length of a static string.

looks like _Lengthof is now _Countof?  again, sounds like the standard isn't settled yet.  firmly against at this time.

**Comment by alejandro-colomar** [comment]
> looks like you've retconned the name of the issue here, and your ask. sounds like the standard isn't settled yet, in which case i'd be firmly against taking any action at this time.

The standard wasn't settled last year when I opened this, which is why I was gathering opinion.

The standard is settled now, and it will remain as `_Countof` for C2y.

> 
> what is the use of an int[42][0] ? seems like this use case was concocted to justify its own existence.

No, that wasn't even brought to the proposal by which `_Countof` was standardized.  `int [42][0]` is something I realized recently, when some Bell Labs known person asked why we don't have arrays of zero elements in ISO C, and suggested that we standardize it.

As to why arrays of 0 elements are useful: they allow generalizing algorithms that might also work with zero elements.  It is more common to have these as dynamic arrays through malloc(0), but the idea with
arrays of 0 elements is essentially the same.

> not sure what is "quite dangerous" about a compile time macro giving the length of a static string.

Off-by-one errors creeping in.  At least if it worked for both arrays and strings.

It would be fine to have

```c
#define STRLEN(s)  (_Countof("" s "") - 1)
```

But overload `_Countof` would be really dangerous.

> looks like _Lengthof is now _Countof? again, sounds like the standard isn't settled yet. firmly against at this time.

_Lengthof is now _Countof, and it has settled that way.
